implementations/rust/ockam/ockam_core/src/compat.rs

```diff
@@ -0,0 +1,178 @@
+//! A facade around the various collections and primitives needed
+//! when using "std", "no_std + alloc" or "no_std" targets.
+//!
+//! When importing from the standard library:
+//!
+//!   1. always prefer core::<mod> over std::<mod> where it's
+//!      available. (e.g. std::fmt::Result -> core::fmt::Result)
+//!   2. use ockam_core::compat::<mod> equivalents where
+//!      possible. (e.g. std::sync::Arc -> ockam_core::compat::sync::Arc)
+//!   3. if you need to add new items to compat, follow the originating
+//!      namespace. (e.g. compat::vec::Vec and not compat::Vec)
+
+/// std::borrow
+#[cfg(all(not(feature = "std"), feature = "alloc"))]
+pub use alloc::borrow;
+#[cfg(feature = "std")]
+pub use std::borrow;
+
+/// std::boxed
+pub mod boxed {
+    #[cfg(all(not(feature = "std"), feature = "alloc"))]
+    pub use alloc::boxed::Box;
+    #[cfg(feature = "std")]
+    pub use std::boxed::Box;
+}
+
+/// std::collections
+pub mod collections {
+    #[cfg(all(not(feature = "std"), feature = "alloc"))]
+    pub use alloc::collections::{BTreeMap, BTreeSet, BinaryHeap, LinkedList, VecDeque};
+    #[cfg(feature = "std")]
+    pub use std::collections::{BTreeMap, BTreeSet, BinaryHeap, LinkedList, VecDeque};
+
+    pub use hashbrown::{HashMap, HashSet};
+}
+
+/// std::error::Error trait
+pub mod error {
+    #[cfg(not(feature = "std"))]
+    pub trait Error: core::fmt::Debug + core::fmt::Display {
+        fn source(&self) -> Option<&(dyn Error + 'static)> {
+            None
+        }
+    }
+    #[cfg(feature = "std")]
+    pub use std::error::Error;
+}
+
+/// std::format
+#[cfg(all(not(feature = "std"), feature = "alloc"))]
+pub use alloc::format;
+#[cfg(feature = "std")]
+pub use std::format;
+
+/// std::io
+#[cfg(not(feature = "std"))]
+pub use core2::io;
+#[cfg(feature = "std")]
+pub use std::io;
+
+/// std::net
+#[cfg(feature = "std")]
+pub use std::net;
+
+/// rand
+pub mod rand {
+    pub use rand::distributions;
+    pub use rand::prelude;
+    pub use rand::CryptoRng;
+    pub use rand::Error;
+    pub use rand::Rng;
+    pub use rand::RngCore;
+
+    #[cfg(not(feature = "std"))]
+    pub use not_random::thread_rng;
+    #[cfg(feature = "std")]
+    pub use rand::thread_rng;
+
+    #[cfg(not(feature = "std"))]
+    pub use not_random::random;
+    #[cfg(feature = "std")]
+    pub use rand::random;
+
+    /// rngs
+    #[cfg(feature = "std")]
+    pub use rand::rngs;
+    #[cfg(not(feature = "std"))]
+    pub mod rngs {
+        pub use super::not_random::OsRng;
+    }
+
+    /// Placeholders for various features from 'rand' that are not
+    /// supported on no_std targets.
+    ///
+    /// WARNING: Thess implementations are NOT random, please do not
+    /// try to use these in production!
+    #[cfg(not(feature = "std"))]
+    mod not_random {
```

Comment:
Why is it needed?

Reply:
`no_std` lacks some of the operating system based functions. So those are placeholders so we can carry on developing until we have implementations for the various hardware we'll be supporting.

Reply:
Let's make sure we do everything to prevent this implementation from been used. If compilation error here is stopping you from continuing your work, how about runtime panic? Or at least warning in the log?

Reply:
`panic!` is no good because I do need to actually run them. Logging also will be very verbose. 

That said, these currently only run for `no_std` and we'll be replacing them with their hardware-specific equivalents before we release the dev preview.

Maybe hide them behind a feature for now like `unsafe_random`?

Reply:
Yeah, would be great. Maybe compiler warning also works

Reply:
Cool, I've added the feature, it's disabled by default. Also looks like `compile_warning!` is still not a thing. :-P